[자동차 경주 구현] 모찌 미션 제출합니다.

wondroid-world opened the pull request:
안녕하세요.
리뷰어(님...?)!! 저는 안드로이드 7기를 하고 있는 모찌입니다.
만나 뵙게되어서 너무 반갑습니다!

페어프로그래밍을 하면서 궁금한 부분이 생겨서 글을 남겨놨습니다.
P.S. 저의 리뷰어가 되어주어서 너무나 감사하고 리뷰어님의 귀한 시간을 내주어서 너무너무 감사해요! 오늘도 포근한 하루 되세용~*

1. validation을 어디에 둬야하는 지 생각을 해봤습니다.
 
 저는 원래 validate를 inputView에서 값을 받고 Controller에서 validate 검증을 하고 객체에 값을 넣었습니다. 하지만, 이번에 오이와 페어 프로그래밍을 하면서, 객체를 만들 때, init에 validate를 넣어서 처리하는 방법을 알게 되었습니다.
 
 그래서, 각각 어떠한 장점이 있는 지 조사했습니다.
 오이의 방법(객체를 생성할 때, validate를 실행)
 - 객체의 안정성과 일관성 보장
 
 외부에서 validation을 진행
 - 불필요한 객체 생성을 맞음
 - 유연한 예외 처리와 성능 최적화
 
 이렇게 각각의 장점이 있는 것을 발견했습니다. 
 
 저의 최종적인 생각은 차 이름과 시도 횟수를 console 창에서 받아올 때, blank, empty 여부 등은 외부에서 진행을 하는 것이 좋다고 생각합니다.
 
 그리고 자동차 이름 길이와 같은 객체의 속성에 중요한 부분은 객체 내부에서 validate를 진행하는 것이 성능 최적화와 객체의 안전성을 보장할 수 있다고 생각합니다. 
 
 이 이부분에서 reviewer는 어떻게 생각하고 있는지 궁금합니다.

2. data class로 Car 클래스를 정의할 때, position은 var vs val 중 무엇을 사용해야할까?

사실 이 부분은 아직도 고민을 하고 있는 부분입니다. 저는 개인적으로는 게임과 같이 속성의 값이 계속 변한다면, position을 val로 지정하여, copy()하는 것 보다 var을 해서 객체의 속성을 바꾸는 것을 선호했습니다.

하지만, 크루들을 이야기를 동시성 문제와 객체의 불변성을 유지하기 위해서 val로 생성해야한다고 이야기를 들었습니다. 

이 부분은 프로덕트의 성격과 전체적이 규모를 생각해서 정해야하는 부분이라고 생각합니다. 궁금한 점이 저희처럼 초기에 작성을 하여 규모를 점차 키워가는 입장이라면, var와 val 중 어떠한 부분을 선택하실 지 궁금합니다. 

긴글 읽어주어서 감사합니다. ❤️

cucumber99 committed:
docs: 기능 구현 목록

cucumber99 committed:
feat: 자동차 객체 구현

cucumber99 committed:
feat: 자동차 리스트 객체 구현

cucumber99 committed:
feat: 숫자 생성 인터페이스 구현

cucumber99 committed:
feat: 난수 생성 기능 구현

cucumber99 committed:
feat: 자동차 경주 기능 구현

cucumber99 committed:
fix: 유효성 검사 로직 오류 수정

cucumber99 committed:
feat: 시도 횟수 객체 구현

cucumber99 committed:
feat: 입력 뷰 구현

cucumber99 committed:
feat: 출력 뷰 구현

cucumber99 committed:
feat: 뷰 컨트롤러 구현

cucumber99 committed:
feat: 컨트롤러 구현

cucumber99 committed:
feat: 전체 기능 구현

cucumber99 committed:
test: 단위 테스트 코드 작성

cucumber99 committed:
refactor: 코드 컨벤션에 따른 수정

cucumber99 committed:
test: 난수 생성 테스트

cucumber99 committed:
docs: 기능 구현 사항 작성

lee-ji-hoon commented on src/main/kotlin/model/Car.kt:
```diff
@@ -0,0 +1,28 @@
+    val name: String,
+    private var _position: Int = INIT_POSITION,
+) {
+    init {
+        validateCar(name)
+    }
```
> validation을 어디에 둬야하는 지 생각을 해봤습니다.

`Car` 라는 것에만 집중을 해볼까요? `Car`는 현실세계로 따지면 `자동차` 입니다. 

자동차를 만드는 공장에서 해당 자동차의 이름을 확인을 할까요? `확인`을 하는 행동이 아닌 이미 검수가 다 거쳐져서 `이름`을 자동차에 추가를 할 뿐이라고 생각을 해요.

그러면 `Car` 라는 객체가 해야 하는 역할은 무엇일까?를 페어 분이랑 고민을 해보시고 결정을 내려보시면 좋을거 같습니다.

lee-ji-hoon replied:
추가로 전반적으로 `init` 을 많이 사용해주시고 계신데 `init`을 사용했을 때 단점은 없을까요? 한 번 고민해보면 좋습니다.

`init` 블럭을 테스트 하려면 어떻게 해야 할까요?

wondroid-world replied:
@lee-ji-hoon 코니! 저도 프리코스할 때, 코니랑 같은 입장이였습니다. 

하지만, 페어와 이야기 하면서 달라졌습니다. Car 객체 초기화할때, 자동차 이름을 검사를 해야한다는 입장으로 바뀌었습니다. 

왜냐하면 input에서 검사를 한번하고 controller에서 모델을 생성한다고 가정하겠습니다. 예로 해킹을 통하여, 이 사이에 데이터가 오염이 될 수 있다고 해서 객체 초기화때, 이름을 한번더 검사해야겠다고 생각했습니다. controller에서 객체를 생성할 때, 늘 옳은 값을 줄지 않을 수도 있다...?라는 걱정도 이유였습니다.

이 부분은 제가 너무 과도한 생각일까요?

lee-ji-hoon replied:
네 모찌가 어떤 의도로 객체를 초기화할 때 자동차 이름을 검사를 해야 한다는 것인지 이해를 했습니다~!

그런데 `Car` 가 만약 다른 공장에서는 Car의 네이밍이 2자도 허용한다고 했을 때는 어떻게 구현을 하게 될까요? 🤔

wondroid-world replied:
음... 코니!

코니의 입장은 자동차의 이름은 외부에서 부여하는 거지, 자동차 자기 자신에게 중요한 부분이 아니다! 그러니, name과 같은 부분은 외부에서 validate하고 객체를 생성해야한다라는 입장인거죠?

SRP원칙에 따라 자동차에게 중요한건 이름이 아니라, 움직이는 거다! 이런 말인건가요?!!!
저의 이해가 맞을까요?

lee-ji-hoon replied:
@wondroid-world  제 생각에는 `SRP` 보다는 이 `Car`라는 객체의 내부에서 특정 validate라는 도메인적인 내용을 직접적으로 의존하고 있는 것에 대해서 고민을 해보면 좋을거 같습니다.

wondroid-world replied:
코니의 말은 재사용을 할 수 있도록 객체를 만들어야한다는 이야기일까요? 
음...! 그럼 Car class에서 이름 검증을 하는 것은 불필요한 부분이라고 생각이 드네용!

오! 제가 너무 검증에 몰두했나봐요! 객체지향 코드를 다시한번 더 생각하게 refresh시켜주어 감사합니다!

lee-ji-hoon commented on src/main/kotlin/controller/UserInterface.kt:
```diff
@@ -0,0 +1,39 @@
+import view.InputView
+import view.OutputView
+
+class UserInterface(
+    private val inputView: InputView = InputView(),
+    private val outputView: OutputView = OutputView(),
+) {
```
`UserInteface` 라는 네이밍을 봤을 때 제가 바로 떠오른건 `User` 들을 정의 하는 Interface인가? 하는 생각이 들었는데요. 조금 더 명확하게 네이밍을 지어보는건 어떨까요?

wondroid-world replied:
코니, 네이밍은 너무 어려운거 같아요. 😭 

먼저, 제가 저렇게 지은 의미는 "사용자의 입출력을 관리하는 controller이다"입니다.

코니의 말대로 저의 의미가 전달이 잘 안된다면, UserInteractionController나 UiFlowController라고 지으면 의미가 잘 전달이 될까요? 코니만의 네이밍 팁이 있을까요?

lee-ji-hoon replied:
`UserInteractionController` 도 나쁘지 않네요! 저도 네이밍이 가장 힘들긴한데요 😭 모찌가 이야기 한 `"사용자의 입출력을 관리하는 controller이다"` 에서 기존 `UserController` 가 입출력과 사용자 중 어떤 것과 관련이 있었을까요? 저는 `사용자` 였던거 같아요. 실제로는 `입출력`이 더 중요한데 말이죠! 

그러니 클래스 명을 지을때도 결국 역할이 무엇이냐에 따라 달라지기에 계속 클래스의 역할에 대해서 집중을 하다 보면 조금이나마 더 쉽게 네이밍을 하게 될 것 같네요!

lee-ji-hoon commented on src/main/kotlin/model/Car.kt:
```diff
@@ -0,0 +1,28 @@
+package model
+
+class Car(
+    val name: String,
+    private var _position: Int = INIT_POSITION,
```
> data class로 Car 클래스를 정의할 때, position은 var vs val 중 무엇을 사용해야할까?

`속성의 값이 계속 변한다면` 이 측면으로 봤을 때 해당 객체는 계속 접근이 된다는 의미입니다. 
그런 상황에서 만약 동시성으로 문제가 생기게 되면 어디서 문제인지 파악이 어려워지게 됩니다.

data class의 생성자로 var을 썼을 때 문제가 될만한 것도 있는데요 이 부분은 한 번 학습 후 고민해보시면 좋을거 같습니다.

wondroid-world replied:
제가 이해하고 공부한 내용은 아래와 같습니다.

프로젝트 특성에 따라서, 멀티스레드이면 val로 프로퍼티를 정의하여, copy()를 한다. 이를 통해, 동시성 문제를 해결한다. 하지만, 싱글 스레드이면 var를 통해서, 메모리 비용의 절감과 성능 최적화를 한다. 

이렇게 보았을 때, 저희의 프로젝트 특성에 따라 사용하면 된다 var또는 val를 사용하면 된다라고 이해했습니다. 제가 이해한 내용이 맞을까요?

wondroid-world replied:
```
data class Car(
    val name: String,
) {
    var position: Int = INIT_POSITION
        private set

    fun moveForward(number: Int) {
        if (number >= MINIMUM_MOVE_POINT) position++
    }

    companion object {
        const val MINIMUM_MOVE_POINT = 4
        const val INIT_POSITION = 0
    }
}
```

고민을 해봤을 때, 현재 저희 프로젝트 상황에서 멀티 스레드까지 사용하지 않을꺼 같아서 성능을 더 높이기 위해서 var을 쓰는 쪽으로 선택했습니다.

lee-ji-hoon replied:
네 좋습니다. 그러면 지금 position이 `private set` 으로 구현이 되어 있는데 `get` 은 public 으로 열어두는 형태 보다 객체에 메세지를 던지는 형태로 구현을 한다면 position 자체에 접근을 막아 볼 수 있을거 같네요 😀

wondroid-world replied:
코니!

코니 말처럼 수정했습니다!

[반영한 커밋](https://github.com/wondroid-world/kotlin-racingcar/commit/7054806fe2a8d1b9f948541655e747ca763658f8)

wondroid-world replied:
코니 저 궁금한 점이 있어요! 

인터넷 커뮤니티에서 회사마다 propert naming하는 규칙이 있기 때문에 객체에서 "."으로 바로 접근하는 것보다 메서드로 접근하는 것이 보안상 안전하다.

위와 같은 글을 읽었습니다. 정말 현직에서도 property에 직접적으로 접근하는 것보다 메서드로 접근하는 이유가 위와 같은 이유가 포함되어 있을까요?

lee-ji-hoon replied:
@wondroid-world 

무조건적으로 property에 접근을 안 하는 것은 아닙니다. 실제로 Android를 많이 다루다 보면 View의 x,y에 그냥 접근을 하기도 하죠.

그런데 그 정보들을 가공을 해서 무언가를 하는 행동은 잘 안하게 되는데요! 그런 관점에서 봤을 때 제가 다시 보니 `position` 은 그대로 private set으로 한 뒤 get property만 열어두는게 나아보이네요!

---

> 즉 제가 잘못 리뷰를 달았던건데요. 제가 원래 달려고 했던건 `Cars`의 `parsedCars` 가 public으로 열려 있고 저 `UserInteractionController` 에서 `cars.parsedCars` 을 직접적으로 접근을 해서 가공을 해서 데이터를 만들고 있는데 이 부분을 리뷰하려다가 잘못 이야기를 꺼내게 된거 같네요 😭

lee-ji-hoon commented on src/main/kotlin/model/AttemptCount.kt:
```diff
@@ -0,0 +1,15 @@
+package model
+
+class AttemptCount(private val input: String) {
+    init {
+        validateAttemptCount()
+    }
```
이 클래스도 2가지의 역할을 갖고 있는 것으로 보입니다. 

1. 현재 시도 횟수
2. 시도 횟수 input의 유효성

`AttemptCount` 의 역할은 무엇일까요 🤔

wondroid-world replied:
코니 말처럼 재활용성이 높은 코드를 작성할려면 validation을 따로 분리하는 게 더 맞다고 생각이 들어요! 그래서 controller에 validation을 따로 예정입니다! class 단일 책임의 원칙에 따라서 attemptCount 객체를 잘 생성만하면 되는 데, 너무 많은 역할을 하고 있는 거 같네요!

lee-ji-hoon replied:
네 좋습니다. 그러면 `AttemptCount` 가 엄청 간소화가 되었는데요. 이 클래스의 역할은 어떤 것이고 필요한 상황일까요? 🤔

wondroid-world replied:
페어와 저의 생각에는 게임에서 시도횟수가 중요한 정보라서 객체에 담아서 보관하고 싶었습니다.

코니의 생각에는 너무 많은 정보를 객체로 만든다고 생각이 드시는 건가요? 

또 지금 생각해보니, "정적 타입으로도 충분히 보관될 수 있는 값이니 클래스로 굳이 안 만들어도 되나?"라는 생각이 드네요!

wondroid-world replied:
[수정한 커밋](https://github.com/wondroid-world/kotlin-racingcar/commit/64ea7823f61b02a26a8b289b386afaf5ef2a47d4)

lee-ji-hoon commented on src/test/kotlin/model/CarTest.kt:
```diff
@@ -0,0 +1,59 @@
+        }
+    }
+
+    @Test
+    @DisplayName("숫자가 4 이상이면 1칸 전진한다.")
+    fun carMovedForwardTest() {
+        car = Car("test")
+        car.moveForward(Car.MINIMUM_MOVE_POINT)
+        assertEquals(car.position, Car.INIT_POSITION + 1)
+    }
```
이 테스트 코드는 `Car. MINIMUM_MOVE_POINT`과 `Car.INIT_POSITION` 에 되게 의존적인데요.

만약 요구사항이 `MINIMUM_MOVE_POINT`이 4가 아닌 6이 됐을 때 이 테스트 코드는 자연스럽게 통과가 되겠죠.

하지만 테스트 코드 `DisplayName` 만 봤을 때 통과를 하는게 맞는 테스트 일까요? 실패해야 하는 테스트인데 통과가 되는 것인지 한 번 고민해봐주세요!

> 다른 테스트 코드들도 한 번 고려해주세요~

lee-ji-hoon replied:
이 부분도 다시 챙겨봐주시면 좋을거 같네요~!

lee-ji-hoon commented on src/test/kotlin/model/CarsTest.kt:
```diff
@@ -0,0 +1,44 @@
+import org.junit.jupiter.api.assertThrows
+
+class CarsTest {
+    private lateinit var cars: Cars
```
이런 형태로 전역으로 초기화 해뒀을 때 각 테스트 별로 테스트 분리가 잘 되어 있는 상태일까요 🤔

만약 A라는 테스트에서 Car을 초기화 후 B라는 테스트에서 Car을 초기화 안 했을 때도 문제가 없을까요?

lee-ji-hoon replied:
여기도 한 번 생각해봐주시면 좋을거 같습니다.

아무래도 객체의 역할에 대해서 집중을 하다 보니 놓치신 리뷰 포인트 같네요!~

lee-ji-hoon commented on src/main/kotlin/view/InputView.kt:
```diff
@@ -0,0 +1,8 @@
+package view
+
+class InputView {
+    fun getUserInput(): String {
+        val input = readln().trim()
+        return input
+    }
+}
```
Input / Output을 나눠주신분 좋습니다! 

그런데 이것도 class일 필요가 있을까요? object여도 되지 않을까 라는 생각이 드는데 class 와 object를 선택하는 방법이 두 분은 어떻게 되실까요?

wondroid-world replied:
사실 제가 여기서 object를 안 쓴 이유는 "object를 잘 모른다."라는 이유였습니다.

지금 공부해보니, object와 class의 큰 차이는 '객체를 몇개 생성할 수 있나?'입니다.
이 관점에서 InputView와 OutputView는 이 프로그램 전체에서 하나만 필요하니, class보다 object로 작성하는 것이 더 바람직해 보입니다!

lee-ji-hoon replied:
네 좋습니다. 그러면 object로 선언을 했을 때 테스트 코드 작성이 어려워서 사용을 하지 않는 사람들도 있는데요. 그럼 왜 `어떤 상황에서 테스트 코드 작성이 어렵다.` 라고 할까요? 

> object 는 1개만 생성이 된다. 가 키워드 입니다!

wondroid-world replied:
코니! 싱글톤이여서 상태가 공유가 되어서 테스트가 어렵다는 이야기일까요?

lee-ji-hoon replied:
네 맞습니다. object 안에서 특정 객체 혹은 상태가 있다면 테스트 할 때 마다 그것을 초기화 해줘야 하는데 기존에 초기화 코드가 없었을 수도 있기 때문에 object 같은 싱글톤이 테스트가 어려운 상황은 특정 상태를 갖고 있을 때입니다.

lee-ji-hoon commented on src/main/kotlin/model/RandomNumberGenerator.kt:
```diff
@@ -0,0 +1,13 @@
+
+import kotlin.random.Random
+
+class RandomNumberGenerator : NumberGenerator {
+    override fun generate(): Int {
+        return Random.nextInt(MAX_VALUE)
+    }
+
+    companion object {
+        const val MAX_VALUE = 9
+    }
+}
```
이 클래스가 하는 일은 `RandomNumber` 를 만드는 것은 맞습니다. 

하지만 이 클래스만 봤을 때 `RandomNumber` 를 만들어주는 클래스구나! 하고 바로 사용을 하게 될거 같은데 막상 보니 `MAX_VALUE = 9`로 선언이 되어있습니다.

그러면 이게 진짜 `RandomNumber`를 만드는게 맞을까요? 이 클래스를 만들어야 했던 이유를 한 번 생각해봅시다!

wondroid-world replied:
맞아요! 코니의 말처럼 재사용성을 위해서 maxValue는 외부에서 받아오는 게 맞다고 생각이 들어서 아래와 같이 수정했습니다!

```
package racingCar.model

import kotlin.random.Random

class RandomNumberGenerator(private val maxValue: Int) : NumberGenerator {
    override fun generate(): Int {
        return Random.nextInt(maxValue)
    }
}

```

lee-ji-hoon replied:
그러면 궁금한게 이것을 클래스까지 빼서 했을 때의 이점은 어떤 것일까요? 

단순 사용하는 곳에서 `Random.nextInt(maxValue)` 으로 사용해도 충분히 보는 입장에서 충분히 이해가 될거 같은데 모찌의 의견이 더 알고 싶네요~!

lee-ji-hoon commented on src/test/kotlin/model/AttemptCountTest.kt:
```diff
@@ -0,0 +1,44 @@
+class AttemptCountTest {
+    private lateinit var attemptCount: AttemptCount
+
+    @Test
+    @DisplayName("시도횟수가 입력되지 않으면 예외가 발생한다.")
+    fun attemptCountNullTest() {
+        assertThrows<IllegalArgumentException> {
+            attemptCount = AttemptCount("")
+        }
+    }
```
`입력이 되지 않으면` 보다는 `빈 값이 입력이 되면` 이 좀 더 맞는 내용 같네요!

lee-ji-hoon commented on src/test/kotlin/model/AttemptCountTest.kt:
```diff
@@ -0,0 +1,44 @@
+        }
+    }
+
+    @Test
+    @DisplayName("숫자가 아닌 값이 입력되면 예외가 발생한다.")
+    fun attemptCountNotIntegerTest() {
+        assertThrows<IllegalArgumentException> {
+            attemptCount = AttemptCount("1A")
+        }
+    }
+
+    @Test
+    @DisplayName("시도횟수가 1미만이면 예외가 발생한다.")
+    fun attemptCountLessThanOneTest() {
+        assertThrows<IllegalArgumentException> {
+            attemptCount = AttemptCount("0")
+        }
+        assertThrows<IllegalArgumentException> {
+            attemptCount = AttemptCount("-1")
+        }
+    }
```
전반적으로 `양수가 아니면 에러가 발생한다.` 라고 귀결이 되는데요

`@ParameterizedTest` 를 사용하는 방식으로 한 번에 처리가 가능해보이네요!

lee-ji-hoon commented on src/test/kotlin/model/RaceTest.kt:
```diff
@@ -0,0 +1,62 @@
+    private lateinit var race: Race
+    private lateinit var cars: Cars
+
+    @BeforeEach
+    fun setup() {
+        cars = Cars("test1,test2")
+    }
```
`Race`를 테스트 할 때 만약 4개의 Car를 테스트 해야 하는 상황이 생기면 여기에 2개를 더 추가를 해야 할텐데 그럼 다른 테스트 코드의 영향이 없을까요 🤔

wondroid-world replied:
코니의 말은 beforeEach를 쓰는 것보다 각 함수에서 초기화를 해주는 것이 테스트의 다양성을 줄 수 있다는 의견인걸까요? 그 의견이라면 저도 동의합니다. 그래서, 이번에 작성하는 테스트 코드는 각 테스트에서 초기화를 해주고 있습니다

lee-ji-hoon commented on src/main/kotlin/model/Cars.kt:
```diff
@@ -0,0 +1,21 @@
+package model
+
+class Cars(private val input: String) {
+    lateinit var parsedCars: List<Car>
```
전반적으로 `lateinit var` 키워드를 사용해주셨는데요! 사용에 문제가 있는 것은 아니지만 `lateinit var` 을 사용 했을 때 주의해야 할 점이 있고 그것을 방지하기 위한 방법도 있는데요! 어떤 것일까요?!

lee-ji-hoon commented on README.md:
```diff
@@ -1 +1,84 @@
+    </tr>
+</table>
+
+## 기능 구현 목록
+
+### 자동차
```
`기능 요구 사항` 을 기반으로 `기능 구현 목록` 을 만들게 될텐데요!

이것을 커밋 단위처럼 나누고 체크 박스로 해서 하나씩 체크해보는 것을 도전해보면 어떨까요?

처음에는 되게 어색하고 중간에 내가 생각했던 것 보다 더 많은 체크 포인트가 생길수도 있지만 이런 체크포인트가 결국 설계를 하는 습관으로 이루어지기에 한 번 도전해보시는 것을 추천드립니다!

wondroid-world replied:
다음부터 꼭 실천하겠습니다. 조언감사합니다!

lee-ji-hoon commented on README.md:
```diff
@@ -1 +1,84 @@
+- 출력
+    - [x] 라운드 별 각 자동차의 위치
+    - [x] 최종 우승자
+## 🚫 예외 상황
+<table>
+   <tr>
+      <td>대상</td>
+      <td>상황</td>
+   </tr>
```
예외상황을 잘 나눠주셨네요 👍

lee-ji-hoon reviewed (changes requested):
안녕하세요! 전반적으로 잘 읽히는 코드였습니다.

하지만 첫 과제인 만큼 `객체의 역할`에 대해서 집중해보면 좋을거 같아서 몇몇 리뷰를 남겨놨으니 고민 부탁드립니다!

고민해보시고 궁금한 점은 편하게 메세지 혹은 코멘트로 남겨주세요 👍

wondroid-world committed:
refactor(InputView, OutputView): class -> object로 수정

- class -> object 변경 사유
- 전체 프로그램 중 하나의 객체만 필요함
- 객체의 상태 관리가 필요 없음

wondroid-world committed:
refactor(UserInteractionController): controller name 변경과 InputView와 OutputView 객체로 변경

- InputView와 OutputView를 class -> object로 변경하여, 기존 UserInface에서 property로 있는 거 삭제
- UserInterface에서 UserInteractionController로 class name 변경

wondroid-world committed:
refactor(Car): Car 클래스에서 검증 단계 삭제

- Model에서 검증이 아닌 Controller 단계에서 검증
- 단일 책임의 원칙에 따라 검증 단계는 분리

wondroid-world committed:
refactor(Cars): 검증 삭제 & 자동차 경주 담당

- Cars 이름 검증 부분 controller로 분리
- 클래스의 확장성을 고려해 Race class 대신 Cars class에서 경주 진행

wondroid-world committed:
refactor: 폴더 구조 변경

wondroid-world committed:
refactor(AttemptCount): 검증 부분 삭제

wondroid-world committed:
refactor(Cars): 이름외 공백 제거

wondroid-world committed:
refactor(RacintController): validation 과정 추가

wondroid-world committed:
refactor(RandomNumber): 재사용성을 위해, maxValue 추출

wondroid-world committed:
refactor(Cars): 확장성을 위해, maxValue 생성자 추가

wondroid-world commented:
@lee-ji-hoon 

코니! 코니 덕분에 확장성과 재사용성에 대해서 고민하게되었습니다!
코니 말을 듣고 저의 코드를 보는 데, 확실히 모듈화가 되어있지 않아서 저의 코드를 다른 곳에서 쓸 수 없더라구요! 앞으로 어떻게 더 SRP와 확장성, 재사용성에 대해서 생각해야겠어요! 감사합니다! 👍

wondroid-world committed:
Merge branch 'step1' into wondroid-world

lee-ji-hoon commented on src/main/kotlin/model/Car.kt:
```diff
@@ -1,27 +1,16 @@
 package model
 
-class Car(
+data class Car(
```
class -> data class 가 되었네요~! `data class` 는 일반 `class` 와 무엇이 다르며, 그 다른 점으로 어떤 상황에서 이점이 있을까요?

wondroid-world replied:
data class로 정의되어 있으니, toString, copy, hashCode를 자동으로 생성해주는 이점이 있습니다.

제가 배우기에는 데이터를 넘길 때 특화된 class이다. 라고 생각합니다. 

제가 백엔드 개발자이면, Car의 name과 position 정보를 DB에 저장해야하는 데이터를 넘길 때, 특화된 data class를 사용할거 같습니다!

wondroid-world replied:
지금 생각해보면 class와 data class를 크게 구분하게 되는 이유는 equals때문이라고 생각합니다. 각 객체가 동등한지 비교할 수 있다는 점에서 data class를 사용했습니다.

lee-ji-hoon commented on src/main/kotlin/model/Car.kt:
```diff
@@ -1,27 +1,16 @@
-        require(name.trim().isNotEmpty()) { "자동차 이름이 올바르지 않습니다." }
-        require(name.length <= MAX_NAME_LENGTH) { "자동차 이름은 5자를 초과할 수 없습니다." }
+        if (number >= MINIMUM_MOVE_POINT) position++
     }
```
만약 Race에서 `MINIMUM_MOVE_POINT` 가 4가 아닌 2로 변경이 된다고 하면 Car라는 클래스를 어떻게 변경 해야 할까요? 아니면 Car가 변경이 되는게 맞을까요?

저는 `MINIMUM_MOVE_POINT` 은 `경기의 규칙` 이라고 보는데 모찌도 한 번 생각해보면 좋을거 같습니다.

wondroid-world replied:
코니의 말 처럼 경기의 규칙에 종속되어서가 아닌 Car는 자동차 생성하는 게 중요하다고 생각해서 아래와 같이 Car Class를 조금더 간단하게 수정했습니다.

```
package racingCar.model

data class Car(
    val name: String,
    private var position: Int = INIT_POSITION,
) {
    fun getPosition() = position

    fun moveForward() = position++

    companion object {
        const val INIT_POSITION = 0
    }
}

```

lee-ji-hoon reviewed (approved):
모찌 안녕하세요~! 객체의 역할에 대해서 심도 있게 고민해주신 부분들 잘 확인했습니다!

이번 미션은 이정도 고민에서 끝내고 다음 리팩토링을 진행해보면 좋을거 같아서 머즈 진행하겠습니다.

지금 제가 추가로 남긴 리뷰나 테스트 코드 리뷰 중에서 일부 반영이 안 되어 있는 부분들이 있는데 이것들은 앞으로의 미션에서 고민해보고 모찌만의 정답을 내려서 반영 혹은 미반영해도 괜찮습니다.
